Advanced Backup Setting Arguments backup_options is not optional

### Terraform CLI and Terraform AWS Provider Version

Terraform v0.14.7

### Affected Resource(s)

[aws_backup_plan](https://registry.terraform.io/providers/hashicorp/aws/3.28.0/docs/resources/backup_plan)

### Terraform Configuration Files

```hcl
resource "aws_backup_plan" "example" {
  name = "tf_example_backup_plan"

  rule {
    rule_name         = "tf_example_backup_rule"
    target_vault_name = aws_backup_vault.test.name
    schedule          = "cron(0 12 * * ? *)"
  }

  advanced_backup_setting {
    resource_type = "EC2"
  }
}
```

### Debug Output

```
Error: error creating Backup Plan: MissingParameterValueException: Resource Type or Backup Options is null in Advanced Backup Settings
{
  RespMetadata: {
    StatusCode: 400,
    RequestID: "1a493370-ed04-4ca7-93d3-14d58e2c8862"
  },
  Code_: "ERROR_4214",
  Message_: "Resource Type or Backup Options is null in Advanced Backup Settings"
}
```

### Expected Behavior

Creation of aws_backup_plan resource without stating the `backup_options` argument.

### Actual Behavior

`tf plan` produces the error that states that the `backup_options` argument inside `advanced_backup_setting` is required, which is unlike what is stated in the documentation:

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/backup_plan#advanced-backup-setting-arguments

But in the latest docs of CloudFormation it states that the `BackupOptions` argument IS required: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-backup-backupplan-advancedbackupsettingresourcetype.html

### Steps to Reproduce

1. `terraform apply`

@lassebenni Thanks for raising this issue.
Indeed, the [AWS Backup API Reference](https://docs.aws.amazon.com/aws-backup/latest/devguide/API_AdvancedBackupSetting.html) states backup options are optional but I can reproduce the error you got with a modified acceptance test:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsBackupPlan_AdvancedBackupSetting'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsBackupPlan_AdvancedBackupSetting -timeout 120m
=== RUN   TestAccAwsBackupPlan_AdvancedBackupSetting
=== PAUSE TestAccAwsBackupPlan_AdvancedBackupSetting
=== CONT  TestAccAwsBackupPlan_AdvancedBackupSetting
    resource_aws_backup_plan_test.go:533: Step 1/3 error: Error running apply: exit status 1
        2021/02/18 17:48:26 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.13.0
        
        Error: error creating Backup Plan: MissingParameterValueException: Resource Type or Backup Options is null in Advanced Backup Settings
        {
          RespMetadata: {
            StatusCode: 400,
            RequestID: "aab23569-3c28-4c67-933b-8a02b320b349"
          },
          Code_: "ERROR_4214",
          Message_: "Resource Type or Backup Options is null in Advanced Backup Settings"
        }
        
        
--- FAIL: TestAccAwsBackupPlan_AdvancedBackupSetting (8.17s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	8.264s
FAIL
GNUmakefile:27: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

I'll fix the resource so that `backup_options` and `resource_type` are both `Required`.

--- a/.changelog/17692.txt
+++ b/.changelog/17692.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_backup_plan: `backup_options` and `resource_type` attributes in `advanced_backup_setting` configuration block are both required
+```

--- a/aws/resource_aws_backup_plan.go
+++ b/aws/resource_aws_backup_plan.go
@@ -126,12 +126,12 @@ func resourceAwsBackupPlan() *schema.Resource {
 					Schema: map[string]*schema.Schema{
 						"backup_options": {
 							Type:     schema.TypeMap,
-							Optional: true,
+							Required: true,
 							Elem:     &schema.Schema{Type: schema.TypeString},
 						},
 						"resource_type": {
 							Type:     schema.TypeString,
-							Optional: true,
+							Required: true,
 							ValidateFunc: validation.StringInSlice([]string{
 								"EC2",
 							}, false),
@@ -339,6 +339,12 @@ func expandBackupPlanAdvancedBackupSettings(vAdvancedBackupSettings *schema.Set)
 		}
 		if v, ok := mAdvancedBackupSetting["resource_type"].(string); ok && v != "" {
 			advancedBackupSetting.ResourceType = aws.String(v)
+		}
+
+		// https://github.com/hashicorp/terraform-plugin-sdk/issues/588
+		// Map in Set may add empty element. Ignore it.
+		if advancedBackupSetting.ResourceType == nil {
+			continue
 		}
 
 		advancedBackupSettings = append(advancedBackupSettings, advancedBackupSetting)

--- a/aws/resource_aws_backup_plan_test.go
+++ b/aws/resource_aws_backup_plan_test.go
@@ -553,6 +553,19 @@ func TestAccAwsBackupPlan_AdvancedBackupSetting(t *testing.T) {
 				ImportState:       true,
 				ImportStateVerify: true,
 			},
+			{
+				Config: testAccAwsBackupPlanConfigAdvancedBackupSettingUpdated(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsBackupPlanExists(resourceName, &plan),
+					resource.TestCheckResourceAttr(resourceName, "name", rName),
+					resource.TestCheckResourceAttr(resourceName, "advanced_backup_setting.#", "1"),
+					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "advanced_backup_setting.*", map[string]string{
+						"backup_options.%":          "1",
+						"backup_options.WindowsVSS": "disabled",
+						"resource_type":             "EC2",
+					}),
+				),
+			},
 		},
 	})
 }
@@ -1010,7 +1023,7 @@ resource "aws_backup_plan" "test" {
 func testAccAwsBackupPlanConfigAdvancedBackupSetting(rName string) string {
 	return fmt.Sprintf(`
 resource "aws_backup_vault" "test" {
-  name = "%[1]s-1"
+  name = %[1]q
 }
 
 resource "aws_backup_plan" "test" {
@@ -1031,6 +1044,38 @@ resource "aws_backup_plan" "test" {
     backup_options = {
       WindowsVSS = "enabled"
     }
+
+    resource_type = "EC2"
+  }
+}
+`, rName)
+}
+
+func testAccAwsBackupPlanConfigAdvancedBackupSettingUpdated(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_backup_vault" "test" {
+  name = %[1]q
+}
+
+resource "aws_backup_plan" "test" {
+  name = %[1]q
+
+  rule {
+    rule_name         = %[1]q
+    target_vault_name = aws_backup_vault.test.name
+    schedule          = "cron(0 12 * * ? *)"
+
+    lifecycle {
+      cold_storage_after = 30
+      delete_after       = 180
+    }
+  }
+
+  advanced_backup_setting {
+    backup_options = {
+      WindowsVSS = "disabled"
+    }
+
     resource_type = "EC2"
   }
 }

--- a/website/docs/r/backup_plan.html.markdown
+++ b/website/docs/r/backup_plan.html.markdown
@@ -67,8 +67,8 @@ For **copy_action** the following attributes are supported:
 ### Advanced Backup Setting Arguments
 For `advanced_backup_setting` the following attibutes are supported:
 
-* `backup_options` - (Optional) Specifies the backup option for a selected resource. This option is only available for Windows VSS backup jobs. Set to `{ WindowsVSS = "enabled" }` to enable Windows VSS backup option and create a VSS Windows backup.
-* `resource_type` - (Optional) The type of AWS resource to be backed up. For VSS Windows backups, the only supported resource type is Amazon EC2. Valid values: `EC2`.
+* `backup_options` - (Required) Specifies the backup option for a selected resource. This option is only available for Windows VSS backup jobs. Set to `{ WindowsVSS = "enabled" }` to enable Windows VSS backup option and create a VSS Windows backup.
+* `resource_type` - (Required) The type of AWS resource to be backed up. For VSS Windows backups, the only supported resource type is Amazon EC2. Valid values: `EC2`.
 
 ## Attributes Reference
 

